File: pset/Pset6/solution_by_Yu_credit.py
# TODO
from cs50 import get_int

# Numbers of digits
N = 16

# Get input
while 1:
    numbers = get_int("Number: ")
    if numbers > 50000:
        break

# Get digits
digits = []
for i in range(N):
    digits.append(int(numbers / 10 ** (N - i - 1)))
    numbers -= digits[i] * (10 ** (N - i - 1))

print(digits)

sum_up = 0
# Luhn algorithm
for i in range(N - 1, 0, -2):
    print(digits[i - 1]*2 % 10, end='')
    print(int(digits[i - 1]*2 /10), end='')
    print(digits[i])
    sum_up += digits[i - 1]*2 % 10
    sum_up += int(digits[i - 1]*2 /10)
    sum_up += digits[i]

print(sum_up)
if (sum_up % 10) == 0:
    if (digits[0] == 0 and digits[1] == 3) and (digits[2] == 4 or digits[2] == 7):
        print("AMEX")
    elif (digits[0] == 5) and (digits[1] >= 1 and digits[1] <= 5):
        print("MASTERCARD")
    elif (digits[0] == 4) or (digits[0] == 0 and digits[1] == 0 and digits[2] == 0 and digits[3] == 4):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: pset/Pset1/solution_by_Yu_credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    // Number of digits
    const int N = 16;
    long numbers = 0;
    do
    {
        numbers = get_long("Number: ");
    }
    while (numbers < 10000);

    // Get digits
    int digits[N];
    long numbers1 = numbers;
    for (int i = 0; i < N; i++)
    {
        // printf("%li\n", numbers1);
        digits[i] = numbers1 / (long) pow(10, N - 1 - i);
        numbers1 -= digits[i] * (long) pow(10, N - 1 - i);
    }

    // Print out digits
    // for (int i = 0; i < N; i++)
    // {
    //     printf("%f", pow(10, N - 1 -i));
    //     printf("%i   %i\n",i, digits[i]);
    // }

    // Whether it is valid
    // Luhn's Algorithm check
    // First non zero value check

    // Luhn's algorithm, sum of multiplied
    int sum_multiplied = 0;
    int sum_other = 0;
    int sum_up = 0;
    int sum_product_digits = 0;
    for (int i = N - 1; i >= 0 ; i -= 2)
    {
        sum_product_digits = ((digits[i - 1] * 2) % 10) + (int)((digits[i - 1] * 2) / 10);
        sum_multiplied += sum_product_digits;
        sum_other += digits[i];
    }

    sum_up = sum_multiplied + sum_other;

    // debug
    // printf("The remainder is %i\n",sum_up % 10);
    // printf("The sum is %i  %i %i\n",sum_multiplied, sum_other, sum_up);

    // Luhn's algorithm

    if ((sum_up % 10) == 0)
    {
        // Check AMEX
        if (((digits[0] == 0) & (digits[1] == 3) & digits[2] == 4) | ((digits[0] == 0) & (digits[1] == 3) & (digits[2] == 7)))
        {
            printf("AMEX\n");
            return 0;
        }

        // Check MASTERCARD
        if ((digits[0] == 5) & ((digits[1] >= 1) & (digits[1] <= 5)))
        {
            printf("MASTERCARD\n");
            return 0;
        }

        // Check VISA
        if ((digits[0] == 4) | ((digits[0] == 0) & (digits[1] == 0) & (digits[2] == 0) & (digits[3] == 4)))
        {
            printf("VISA\n");
            return 0;
        }
    }
    // Else
    printf("INVALID\n");
    return 0;
}